### DISCUSSION TITLE
Followpoints appear way earlier than they're supposed to

### DISCUSSION THREAD

**Discussion by Flamiii** [ppy/osu#19807] [Q&A] [answered]
This has been something that has been very noticeable for me and I'm fairly certain that this came from the 2022.810.0 update of lazer. If I had to guess I would say that this might be an unintended side effect of https://github.com/ppy/osu/pull/19587, but I'm not sure. This shouldn't be skin specific as I've tested with multiple skins and it happens with all of them.

Skin 1: https://skins.osuck.net/index.php?newsid=863
Skin 2: https://skins.osuck.net/index.php?newsid=166
Map: https://osu.ppy.sh/beatmapsets/408820#osu/887621

Some examples of them appearing way too early:

![osu_2022-08-16_19-25-39](https://user-images.githubusercontent.com/74088315/185008574-267916df-64a8-4741-bacc-76b8d563d4b3.png)
![osu_2022-08-16_19-29-38](https://user-images.githubusercontent.com/74088315/185008615-394a4386-3175-4545-aba9-565732e357e0.png)
![osu_2022-08-16_19-29-49](https://user-images.githubusercontent.com/74088315/185008946-1a06473a-0652-410f-aace-a2cef0762ab1.png)

The bottom followpoint seems bugged to me as it appears and then disappears without being visibly connected to any notes:

![osu_2022-08-16_19-33-48](https://user-images.githubusercontent.com/74088315/185008668-2c8bbfc0-1dac-4400-ba87-7bc509e968f8.png)

**Comment by peppy** [comment]
Have you tested the build before that one you mention to confirm that's where it starts?

**Reply by Flamiii** [reply]
Yes, I just tested 2022.731.1 and it looks normal to me. I tested the same map using the same skins and the followpoints are not appearing abnormally early and I'm not getting any weird flashing followpoints.

**Reply by peppy** [reply]
I can't reproduce this.

![Parallels Desktop 2022-08-25 at 09 20 38](https://user-images.githubusercontent.com/191335/186627143-155bdf4e-8198-4dd5-a90b-589cf277e39c.png)

![osu! 2022-08-25 at 09 20 45](https://user-images.githubusercontent.com/191335/186627177-6a2d9c38-a889-4184-a5aa-87b7a553bc6e.png)


Please provide a video showing the issue occurring.

**Reply by Flamiii** [reply]
Side by side:
https://www.youtube.com/watch?v=Wti_9aeIovY

**Reply by peppy** [reply]
I've gone through frame by frame and fail to see an issue. The only discrepancy is in one case the frame animation proceeds slightly earlier than the others, but this is to be expected as there's no guarantee of frame perfect animation playback.

Going to have to wait for someone with better eyes to figure this one out, I have no idea sorry.

**Comment by NotGumballer91** [comment]
![image](https://user-images.githubusercontent.com/64581009/186870604-8afefd81-17f3-4e29-87ed-7e43cb1887d0.png)

I've noticed that 2022.821.0 is displaying followpoints way before the circle is even visible.
Is it because of some weird preloading or just skin edge case?

**Reply by peppy** [reply]
First guess would be that the animation isn't being reset correctly on reuse of a pooled drawable.

**Comment by sineplusx** [comment]
I also experience this issue when using my skin with animated followpoints. I am currently running version 2022.911.0

In gameplay, some followpoints render way too early without the corresponding hitcircles being visible. The followpoints in my skin are animated in a way that they appear and decay very quickly, so this shouldn't happen. When I rewind a replay or replay a map after experiencing the bug, it always works as expected. Something I also noticed that isn't visible in the video (but in the screenshots), is that if I autoplayed the map and set the speed to 0,5x, in addition to appearing too early, the followpoints appear a second (late) time after they are supposed to be already gone.

Skin: https://mega.nz/folder/geM12TZS#fecFrFAX_crcS_kLevIV-w/file/VOt1jQhD
Map: https://osu.ppy.sh/beatmapsets/1254200#osu/2608366

Screenshots were made with Autoplay at 0,5x ingame speed

Green 2-3 early
![Green 2-3 early](https://user-images.githubusercontent.com/112003827/190501647-d6223687-58f4-465d-bf76-a0dedb9bb9f3.png)

Green 2-3 late
![Green 2-3 late](https://user-images.githubusercontent.com/112003827/190501692-f3ec63c9-e63f-4233-87d2-7b0e26d12c28.png)

Green 3-4 early
![Green 3-4 early](https://user-images.githubusercontent.com/112003827/190501707-57044a10-7344-40c7-aae4-9697b2e46798.png)

Blue 3-4 early
![Blue 3-4 early](https://user-images.githubusercontent.com/112003827/190501715-1010a61c-ac52-48c7-9521-84577b95014c.png)

Initial play with 1,0x ingame speed
https://user-images.githubusercontent.com/112003827/190503191-722e1c47-575b-47f8-890b-0cef74d1f2bc.mp4

After using the "Clearing all caches" option in the debug options
https://user-images.githubusercontent.com/112003827/190503501-fda51d2c-282c-4dc7-9b08-ed6a1590cee5.mp4

**Comment by smoogipoo** [comment, marked as answer]
https://github.com/ppy/osu/issues/20335
